k256/src/ecdsa.rs

```diff
@@ -120,6 +132,89 @@ pub type SigningKey = ecdsa_core::SigningKey<Secp256k1>;
 #[cfg(feature = "ecdsa")]
 pub type VerifyingKey = ecdsa_core::VerifyingKey<Secp256k1>;
 
+#[cfg(all(feature = "ecdsa", feature = "arithmetic"))]
+/// Extension trait for [`VerifyingKey`] for zkvm
+pub trait PowdrVerifyKey {
+    /// Recover a [`VerifyingKey`] from a pre-hashed message and signature.
+    fn powdr_recover_from_prehash(
+        prehash: &[u8],
+        signature: &Signature,
+        recovery_id: RecoveryId,
+    ) -> Result<VerifyingKey, Error>;
+}
+#[cfg(all(feature = "ecdsa", feature = "arithmetic"))]
+#[allow(non_snake_case)]
+impl PowdrVerifyKey for ecdsa_core::VerifyingKey<Secp256k1> {
+    fn powdr_recover_from_prehash(
+        prehash: &[u8],
+        signature: &Signature,
+        recovery_id: RecoveryId,
+    ) -> Result<Self, Error> {
+        let (r, s) = signature.split_scalars();
+        let z = Scalar::reduce(&bits2field::<Secp256k1>(prehash)?);
+
+        let r_bytes = if recovery_id.is_x_reduced() {
+            let repr: FieldBytes<Secp256k1> = r.to_repr();
+            let r_bytes = <U as FieldBytesEncoding<Secp256k1>>::decode_field_bytes(&repr)
+                .checked_add(&ORDER)
+                .into_option()
+                .ok_or_else(Error::new)?;
+
+            <U as FieldBytesEncoding<Secp256k1>>::encode_field_bytes(&r_bytes)
+        } else {
+            r.to_repr()
+        };
+
+        let R = PowdrAffinePoint(
+            <Secp256k1 as CurveArithmetic>::AffinePoint::decompress(
+                &r_bytes,
+                u8::from(recovery_id.is_y_odd()).into(),
+            )
+            .into_option()
+            .ok_or_else(Error::new)?,
+        );
+
+        let r_inv = r.invert().unwrap();
+        let u1 = -(r_inv * z);
+        let u2 = r_inv * *s;
+        let pk =
+            crate::arithmetic::affine_op::lincomb(&[(PowdrAffinePoint::generator(), u1), (R, u2)]);
+        let vk = Self::from_affine(pk.0).expect("public key recovery failed");
```

Comment:
Wait, what is this doing? Shouldn't this function receive `&self` and use this as the verification key, instead of computing it?